File: scripts/dataforseo_onpage_analyze.py
#!/usr/bin/env python3
"""
On-Page-Analyse per DataForSEO API v3 (instant_pages).

Zugangsdaten (wie im DataForSEO-Dashboard unter „SEO Keyword Tool“ / API Access):
  • scripts/dataforseo_config.py  oder  scripts/seo_keyword_tool_config.py
  • oder Umgebungsvariablen DATAFORSEO_LOGIN + DATAFORSEO_PASSWORD
    (Passwort auch als DATAFORSEO_API_KEY möglich – gleiches Secret wie „API Password“)

Beispiel:
  python scripts/dataforseo_onpage_analyze.py
  python scripts/dataforseo_onpage_analyze.py "https://seomuenchen.com/blog/onpage-seo-checkliste"
"""

from __future__ import annotations

import argparse
import base64
import json
import os
import sys
import urllib.error
import urllib.request
from pathlib import Path

SCRIPTS_DIR = Path(__file__).resolve().parent
API_URL = "https://api.dataforseo.com/v3/on_page/instant_pages"


def _load_credentials_from_py_file(path: Path) -> tuple[str, str] | None:
    """Liest DATAFORSEO_LOGIN + DATAFORSEO_PASSWORD oder DATAFORSEO_API_KEY (Passwort)."""
    import importlib.util

    spec = importlib.util.spec_from_file_location(path.stem, path)
    if not spec or not spec.loader:
        return None
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    l = getattr(mod, "DATAFORSEO_LOGIN", None)
    p = getattr(mod, "DATAFORSEO_PASSWORD", None) or getattr(mod, "DATAFORSEO_API_KEY", None)
    if l and p:
        return str(l), str(p)
    return None


def load_credentials() -> tuple[str, str]:
    login = os.environ.get("DATAFORSEO_LOGIN")
    password = os.environ.get("DATAFORSEO_PASSWORD") or os.environ.get("DATAFORSEO_API_KEY")

    custom = os.environ.get("SEO_KEYWORD_TOOL_CONFIG")
    paths = [
        SCRIPTS_DIR / "dataforseo_config.py",
        SCRIPTS_DIR / "seo_keyword_tool_config.py",
    ]
    if custom:
        paths.insert(0, Path(custom).expanduser())

    for config_path in paths:
        if config_path.is_file():
            creds = _load_credentials_from_py_file(config_path)
            if creds:
                return creds

    if login and password:
        return login, password

    print(
        "Keine Zugangsdaten: Lege scripts/seo_keyword_tool_config.py oder "
        "scripts/dataforseo_config.py an (siehe *_config.example.py) oder setze "
        "DATAFORSEO_LOGIN und DATAFORSEO_PASSWORD (bzw. DATAFORSEO_API_KEY).",
        file=sys.stderr,
    )
    sys.exit(1)


def main() -> None:
    parser = argparse.ArgumentParser(description="DataForSEO On-Page Instant Pages")
    parser.add_argument(
        "url",
        nargs="?",
        default="https://seomuenchen.com/blog/onpage-seo-checkliste",
        help="Zu analysierende URL (öffentlich erreichbar)",
    )
    parser.add_argument(
        "--render",
        action="store_true",
        help="Browser-Rendering (teurer, u.a. Core Web Vitals)",
    )
    parser.add_argument(
        "--save",
        action="store_true",
        help="Antwort nach scripts/dataforseo_last_response.json speichern",
    )
    args = parser.parse_args()

    login, password = load_credentials()
    auth = base64.b64encode(f"{login}:{password}".encode()).decode()

    task: dict = {
        "url": args.url,
        "accept_language": "de-DE,de;q=0.9,en;q=0.8",
    }
    if args.render:
        task["enable_browser_rendering"] = True
        task["disable_cookie_popup"] = True

    body = json.dumps([task]).encode("utf-8")

    req = urllib.request.Request(
        API_URL,
        data=body,
        method="POST",
        headers={
            "Authorization": f"Basic {auth}",
            "Content-Type": "application/json",
        },
    )

    try:
        with urllib.request.urlopen(req, timeout=180) as resp:
            raw = resp.read().decode("utf-8")
    except urllib.error.HTTPError as e:
        print(e.read().decode("utf-8", errors="replace"), file=sys.stderr)
        sys.exit(1)

    data = json.loads(raw)

    if data.get("status_code") != 20000:
        print(json.dumps(data, indent=2, ensure_ascii=False))
        sys.exit(1)

    task_info = data["tasks"][0]
    if task_info.get("status_code") != 20000:
        print(json.dumps(task_info, indent=2, ensure_ascii=False))
        sys.exit(1)

    items = task_info.get("result", [{}])[0].get("items") or []
    page = next((i for i in items if i.get("resource_type") == "html"), None)

    if not page or "meta" not in page:
        print("Keine HTML-Metadaten in der Antwort.", file=sys.stderr)
        print(json.dumps(task_info, indent=2, ensure_ascii=False)[:4000])
        sys.exit(1)

    m = page["meta"]
    content = m.get("content") or {}
    checks = m.get("checks") or {}

    def true_flags(d: dict) -> list[str]:
        return [k for k, v in d.items() if v is True]

    print("\n=== DataForSEO On-Page ===")
    print("URL:", args.url)
    print("Kosten (USD):", data.get("cost"))
    print("onpage_score:", m.get("onpage_score"))
    print("\n--- Meta ---")
    print("Title:", m.get("title") or m.get("meta_title"))
    print("Title-Länge:", m.get("title_length"))
    desc = m.get("description") or ""
    print("Description:", (desc[:200] + "…") if len(desc) > 200 else desc)
    print("\n--- Content ---")
    print("Wörter:", content.get("plain_text_word_count"))
    print("Title↔Content:", content.get("title_to_content_consistency"))
    print("Description↔Content:", content.get("description_to_content_consistency"))
    print("Flesch–Kincaid:", content.get("flesch_kincaid_readability_index"))
    print("\n--- Checks (TRUE = Hinweis) ---")
    flags = true_flags(checks)
    if flags:
        for f in flags:
            print(" •", f)
    else:
        print("(keine booleschen Warn-Flags)")

    if args.save:
        out = SCRIPTS_DIR / "dataforseo_last_response.json"
        out.write_text(json.dumps(data, indent=2, ensure_ascii=False), encoding="utf-8")
        print("\nGespeichert:", out)


if __name__ == "__main__":
    main()

File: scripts/dataforseo-onpage-analyze.mjs
/**
 * On-Page-Analyse einer URL über DataForSEO API v3 (instant_pages).
 *
 * Voraussetzung: In .env.local (nicht committen) – Zugangsdaten wie im Dashboard
 * „SEO Keyword Tool“ / API Access:
 *   DATAFORSEO_LOGIN=deine@email.de
 *   DATAFORSEO_PASSWORD=dein_api_secret
 *   (alternativ statt PASSWORD: DATAFORSEO_API_KEY – gleiches Secret)
 *
 * Aufruf:
 *   node scripts/dataforseo-onpage-analyze.mjs
 *   node scripts/dataforseo-onpage-analyze.mjs "https://seomuenchen.com/blog/onpage-seo-checkliste"
 *
 * Hinweis: Jeder Aufruf kostet laut DataForSEO (siehe Pricing). Kein JS-Rendering standardmäßig
 * (günstiger); für SPAs optional --render setzen.
 */

import { readFileSync, existsSync } from 'fs'
import { fileURLToPath } from 'url'
import { dirname, join } from 'path'

const __dirname = dirname(fileURLToPath(import.meta.url))
const root = join(__dirname, '..')

function loadEnvLocal() {
  const p = join(root, '.env.local')
  if (!existsSync(p)) return
  const raw = readFileSync(p, 'utf8')
  for (const line of raw.split('\n')) {
    const m = line.match(/^\s*([A-Za-z_][A-Za-z0-9_]*)\s*=\s*(.*)$/)
    if (!m) continue
    const key = m[1]
    let val = m[2].trim()
    if ((val.startsWith('"') && val.endsWith('"')) || (val.startsWith("'") && val.endsWith("'")))
      val = val.slice(1, -1)
    if (process.env[key] === undefined) process.env[key] = val
  }
}

loadEnvLocal()

const API = 'https://api.dataforseo.com/v3/on_page/instant_pages'

const targetUrl =
  process.argv[2] || 'https://seomuenchen.com/blog/onpage-seo-checkliste'
const useRender = process.argv.includes('--render')

const login = process.env.DATAFORSEO_LOGIN
const password = process.env.DATAFORSEO_PASSWORD || process.env.DATAFORSEO_API_KEY

if (!login || !password) {
  console.error(
    'Fehlende Variablen: DATAFORSEO_LOGIN und DATAFORSEO_PASSWORD (oder DATAFORSEO_API_KEY) in .env.local'
  )
  process.exit(1)
}

const auth = Buffer.from(`${login}:${password}`).toString('base64')

const task = {
  url: targetUrl,
  accept_language: 'de-DE,de;q=0.9,en;q=0.8',
}

if (useRender) {
  task.enable_browser_rendering = true
  task.disable_cookie_popup = true
}

const body = [task]

const res = await fetch(API, {
  method: 'POST',
  headers: {
    Authorization: `Basic ${auth}`,
    'Content-Type': 'application/json',
  },
  body: JSON.stringify(body),
})

const json = await res.json()

if (!res.ok) {
  console.error('HTTP', res.status, JSON.stringify(json, null, 2))
  process.exit(1)
}

const status = json.status_code
if (status !== 20000) {
  console.error('API status_code:', status, json.status_message)
  console.error(JSON.stringify(json, null, 2))
  process.exit(1)
}

const taskResult = json.tasks?.[0]
if (!taskResult || taskResult.status_code !== 20000) {
  console.error('Task-Fehler:', JSON.stringify(taskResult, null, 2))
  process.exit(1)
}

const items = taskResult.result?.[0]?.items
const page = Array.isArray(items) ? items.find((i) => i.resource_type === 'html') : null

if (!page?.meta) {
  console.error('Keine HTML-Item-Daten:', JSON.stringify(taskResult.result, null, 2).slice(0, 2000))
  process.exit(1)
}

const m = page.meta
const content = m.content || {}
const checks = m.checks || {}
const cost = json.cost ?? taskResult.cost

/** @param {Record<string, boolean>} obj */
function listTrue(obj) {
  return Object.entries(obj)
    .filter(([, v]) => v === true)
    .map(([k]) => k)
}

console.log('\n=== DataForSEO On-Page: ' + targetUrl + ' ===\n')
console.log('Kosten (USD, laut API):', cost)
console.log('onpage_score (0–100):', m.onpage_score ?? '—')
console.log('\n--- Meta ---')
console.log('Title:', m.title ?? m.meta_title)
console.log('Title-Länge:', m.title_length)
console.log('Meta Description:', (m.description || '').slice(0, 160) + (m.description?.length > 160 ? '…' : ''))
console.log('Description-Länge:', m.description_length)

console.log('\n--- Content-Metriken ---')
console.log('Wörter (plain_text_word_count):', content.plain_text_word_count)
console.log('plain_text_rate:', content.plain_text_rate)
console.log('Title ↔ Content Konsistenz (0–1):', content.title_to_content_consistency)
console.log('Description ↔ Content Konsistenz (0–1):', content.description_to_content_consistency)
console.log('Flesch–Kincaid Lesbarkeit:', content.flesch_kincaid_readability_index)

console.log('\n--- Struktur ---')
console.log('internal_links_count:', m.internal_links_count)
console.log('external_links_count:', m.external_links_count)
console.log('images_count:', m.images_count)
if (m.htags) console.log('H-Tags:', JSON.stringify(m.htags, null, 2))

console.log('\n--- Checks (nur TRUE = Problem laut Schwellen) ---')
const problems = listTrue(checks)
if (problems.length === 0) {
  console.log('Keine markierten Check-Flags (oder keine boolean warnings).')
} else {
  problems.forEach((p) => console.log(' •', p))
}

if (m.page_timing?.largest_contentful_paint != null) {
  console.log('\n--- CWV (falls mit Rendering geliefert) ---')
  console.log('LCP ms:', m.page_timing.largest_contentful_paint)
  console.log('CLS:', m.cumulative_layout_shift)
}

console.log('\n--- Kurzinterpretation ---')
const score = m.onpage_score
if (typeof score === 'number') {
  if (score >= 85) console.log('On-Page-Score: stark (' + score + '/100).')
  else if (score >= 65) console.log('On-Page-Score: solide (' + score + '/100), Details bei den Checks prüfen.')
  else console.log('On-Page-Score: Nachholbedarf (' + score + '/100).')
}
if (content.title_to_content_consistency != null && content.title_to_content_consistency < 0.3) {
  console.log('Hinweis: Title passt schwach zum sichtbaren Text (Konsistenz niedrig).')
}
if (content.description_to_content_consistency != null && content.description_to_content_consistency < 0.2) {
  console.log('Hinweis: Meta-Description passt schwach zum Inhalt — prüfen.')
}

console.log('\nVollständige Rohantwort: scripts/dataforseo-last-response.json (optional speichern mit --save)\n')

if (process.argv.includes('--save')) {
  const { writeFileSync } = await import('fs')
  writeFileSync(join(root, 'scripts/dataforseo-last-response.json'), JSON.stringify(json, null, 2), 'utf8')
  console.log('Gespeichert: scripts/dataforseo-last-response.json')
}
